### ISSUE TITLE
NodeJs app in GitHub Codespaces missing environment variable with hyphen

### ISSUE THREAD

**Issue by rudiv** [dotnet/aspire#6680]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

We have a fairly simple Nodejs app spun up alongside SQL Server, Redis, Azure Storage and a .NET API. This works locally (across Windows, Linux and OSX developers), however does not work in GitHub Codespaces.

Our Redis cache is called "project-cache", which in turn tries to set the environment variable `ConnectionStrings__project-cache`. I've also set 2 manual environment variables to confirm the root cause:

![Image](https://github.com/user-attachments/assets/69858bd7-bbc9-436d-9a41-af4c258d511c)

Of the 3, only `ConnectionStrings__project_cache_test` is set.

On Codespaces using the default environment, it seems that DCP is not sending the Environment Variables correctly which causes this to be dropped. As it's a black box I'm unable to see exactly where or why this would happen.

The only solution currently is to remove the hyphen from the resource name within the AppHost.

I know that this is a "Codespaces thing" but it seems that Environment Variables should be reliably set across all platforms / shells, which I'm assuming means that DCP isn't quoting them specifically in the NodeJS application type, as these Environment Variables **are** correctly passed to the .NET application.

### Expected Behavior

Environment variables with hyphens in the name are set.

### Steps To Reproduce

Add an Npm app to the AppHost with an environment variable containing a dash.

```csharp
builder.AddNpmApp("anything", "path/to/anything", "dev")
    .WithEnvironment("Something-Has-A-Hypen", "Anything");
```

Run Aspire within Github Codespaces, note that the environment variable is not set.

### .NET Version info

.NET SDK:
 Version:           9.0.100
 Commit:            59db016f11
 Workload version:  9.0.100-manifests.3068a692
 MSBuild version:   17.12.7+5b8665660

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /usr/share/dotnet/sdk/9.0.100/

.NET workloads installed:
There are no installed workloads to display.
Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0
  Architecture: x64
  Commit:       9d5a6a9aa4

.NET SDKs installed:
  8.0.403 [/usr/share/dotnet/sdk]
  9.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 7.0.20 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.10 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 7.0.20 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.10 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/usr/share/dotnet]

global.json file:
  Not found

### Anything else?

Seems to be a DCP issue. This is running through VS Code with the C# Dev Kit installed on GitHub Codespaces.

**Comment by davidfowl** [comment]
Wild guess that linux doesn't support environment variables with - in the name 😄

**Comment by rudiv** [comment]
Yup, that's not the point of the post though.

Are the .NET projects not also passing configuration through environment variables? If so, it works fine there, just not for the NodeJS apps.

It also works on a Fedora 40 machine, just not in Codespaces (which I'm guessing is an Ubuntu image) that I've tested so far.

You can also use `env` with a quoted string to set "unusual" environment variables, which would let this be set.

`env 'ConnectionStrings__project-cache=localhost:63999' npm run dev` works for example.

**Comment by karolz-ms** [comment]
The only "safe" characters in environment variable names are uppercase letters, digits, and underscore https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html Anything else is implementation-dependent and may or may not be supported. Hyphen in particular is quite problematic as it is used for invocation flags, and different shells on different platforms may parse the command line differently, depending on how things are quoted, whether single- or double hyphens are used for flags etc. etc.

@rudiv would it be OK for you to work around this issue by avoiding the hyphen in env var names?

**Comment by rudiv** [comment]
@karolz-ms yes it's not me that's adding the hyphens, it's aspire based on the resource name. We've obviously just changed the resource names to not have them for now.

My point is that there is a discrepancy between the .NET app and the NodeJS app, assuming that they're both using environment variables (which is what the Aspire Dashboard would imply). The .NET app works, the NodeJS app does not work.

It's my opinion that either Aspire should be aware of this and map the environment variables correctly in one way or another, either by calling `env` on platforms where it's available (less compatible), or adjusting the names to be globally compatible by removing these "special" characters cross platform, and adjusting the internal connection strings/mappings to suit.

**Comment by davidfowl** [comment]
Of it can just fail...

**Comment by rudiv** [comment]
@davidfowl so when code is written on Windows (where the hyphens work) and you deploy to a Linux environment which would attempt to use environment variables, it should just fail? Or a development team that uses different OSes shouldn't be able to collaborate because a simple resource name causes a complete failure cross platform?

Am I just disallowed from making suggestions on this repo?

**Comment by davidfowl** [comment]
Within reason, platforms try to mask the differences as much as possible, it's our job to figure out what is within reason. Sometimes we are able to do it well if it is done at the right level (e.g. maybe we can do it specifically for AddNpmApp). If you told me we should make it work generally across OSes, that'd be a harder pill to swallow.

**Comment by rudiv** [comment]
I fully agree, and I'm just trying to get across that this doesn't work in a specific scenario (`.AddNpmApp()`) whilst still working in others (`.AddProject<>`).

To me, that sounds like there's a difference in some internals somewhere (DCP?) that is doing things differently between targets.

With that said, if the resolution to this is just a note in the documentation that states that resource names with hyphens, whilst allowed by Aspire, may not work across platforms (e.g. linux) with a suggestion to perhaps use `RuntimeInformation.IsOSPlatform` to work around it that would probably suffice. The main issue is that this is undocumented.

**Comment by joperezr** [comment]
[Triage] - we need to do more investigation here to see what breaks so that we can determine whether or not an analyzer is appropriate.

**Comment by mitchdenny** [comment]
@rudiv we took a second look at this. We have codespaces enabled in the dotnet/aspire repo so I was able to test your scenario with the bits on main and we found that it was working fine (you could have a resource called `cache-with-hyphen`, reference it from a Node app, and access that environment variable in the Node code.

Given this I think the next step is having a closer look at your exact Codespaces setup. Is it possible for you to setup a Codespace which replicates the issue you are seeing so that we can take a look?

**Comment by rudiv** [comment]
Sure, [here's a minimal-ish sample](https://github.com/rudiv/aspire-cs). Sorry it's hastily thrown together.

Spinning up a Codespaces on that, running Aspire and hitting port 3000 will show that there's no environment variables set with hyphens (there should be 2).

**Comment by mitchdenny** [comment]
I'm seeing the environment variables come through in the dashboard:

![Image](https://github.com/user-attachments/assets/73d49933-f112-4f5e-9fbe-d045e978df00)

Your vite sample wasn't work so i tweaked it to just be a simple Node script that dumps out the environment:

```js
console.log(JSON.stringify(process.env, null, 2));
```

... and sure enough it dumps out the environment variables:

![Image](https://github.com/user-attachments/assets/ba4e87aa-369b-48ee-8c89-13fcd37d4b36)

Maybe it is vite getting in the road here?

**Comment by rudiv** [comment]
@mitchdenny I don't see `ConnectionStrings__test-i-do-not-work` in your output (though it may be off screen)? The issue is with the hyphens.

**Comment by mitchdenny** [comment]
You are quite right I was looking at the wrong connection string. I managed to reproduce this with plain old `AddExecutable("dump", "/bin/bash", ".", "dump.sh")` where `dump.sh` is just the following:

```bash
#!/bin/bash
export
```

I found that if I reference a resource without hyphens it works, but with hypens it doesn't. @karolz-ms any more thoughts here on what we could do to make this work? It works when you launch a .NET process so there must be a way of doing it.

**Comment by karolz-ms** [comment]
We need to take a closer look

**Comment by paulirwin** [comment]
This affects Linux users that are running Aspire directly as well, not just GitHub Codespaces. The title should probably be updated.

We just encountered this with a developer trying to run our Aspire app on Linux, where we had a Node app referencing a .NET API via `WithReference`, where the resource name of the API had a hyphen.

**Comment by maddymontaquila** [comment]
@mitchdenny I think we maake a call here with @karolz-ms to either convert "-" to "_" in resource names when wee make them env vars, or ban "-" entirely in resource names?
